### ISSUE TITLE
setup jenkins test server

### ISSUE THREAD

**Issue by tillmo** [spechub/Hets#1736]

**Comment by eugenk** [comment]
* Jenkins can run tests for more than 50 minutes (Travis-CI cancels them after 50 minutes). If we add more tests like in #1730 (which we should), we will surpass this limit.
* Jenkins could regularly build Ubuntu-Packages and put them in a PPA.

---

* If it's not too much work, Jenkins should allow creating a build matrix like Travis, i.e. run the tests with different installed software (e.g. different versions of PostgreSQL, Java etc.).

**Comment by jelmd** [comment]
jenkins servers can run more than a single job, so use http://ohj.iks.cs.ovgu.de:8080/ - no need to setup a separate instance.

'regularly build Ubuntu-Packages ...' makes no sense to me. Triggering package makes when a new version got released makes more sense. However, since the intended target is launchpad PPA, it is useless because it builds all packages from source.

The travis problem is IMHO a self-made problem: If you always need to recompile the whole GHC including all required/used modules, one should not wonder, why this takes several days to complete. This wouldn't be different on the jenkins instance if stack would be used there as well ....

Wrt. "If it's not too much work ..." - it would be like being in hell.

**Comment by eugenk** [comment]
> jenkins servers can run more than a single job, so use http://ohj.iks.cs.ovgu.de:8080/ - no need to setup a separate instance.

We can use that, but it had reliability problems with the old Ontohub. It was offline quite a lot (not responding to Github's web hooks) and sometimes the tests were red even though everything was fine. The latter may not be a problem with Hets (we can see this if it happens), but reliability is important.

> 'regularly build Ubuntu-Packages ...' makes no sense to me.

"Regularly" could mean on every release (triggered by a Github web hook when a release/tag is created).

> If you always need to recompile the whole GHC including all required/used modules

This does not happen. They are cached. Only the packages that are added since the previous build need to be compiled. On https://travis-ci.org/spechub/Hets/builds/269109958, this took 5 seconds because there was one new package.

The steps that take "a lot" of time (>= 30 seconds) are
* 58 seconds: Setting up build cache
* 79 seconds: `sudo apt-get install --no-install-recommends `dpkg-checkbuilddeps debian/control 2>&1 | cut -f3- -d":" | sed -e 's,([^)]*),,g' -e 's,openjdk-.*-jdk,,'` || true` Apart from this one, the whole installation is cached.
* 408 seconds: `make`
* 303 seconds: `make check`
* 289 seconds: `test/hets-lib-check.sh`
* 96 seconds: store build cache

In total, restoring and recreating the cache (which includes everything from stack and hets-lib) takes about 2:45 minutes.

> Wrt. "If it's not too much work ..." - it would be like being in hell.

Ok

**Comment by jelmd** [comment]
Wrt. 'reliability': Hmmm, setting up another instance of the same SW (with the same bugs) and hope, that such bugs disappear magically doesn't make any sense to me. If  "un-reliability" gets encountered, one either needs to accept it (usually developers know, how to easily circumvent problems, e.g. by triggering an update manually via e-mail, web, etc.) or take the burden and investigate properly.

E.g. I really doubt, that "It was offline quite a lot", since it is a 24/7 service/server and thus gets usually booted max. once/month after an update - but there are no updates for it anymore, so last 3 boots were in Mar, Jan, Sep). Also a lot of other zones/services are running on this box, so beside temp. resource problems (not enough memory slips into mind, because we have already seen, that hets chomps huge amounts of RAM from time to time) other service customers would have noticed, if there is a general "box problem". Furthermore I'm pretty sure, that git hub notification service is not a reliable but fire-and-forget service. So, when you say "offline", are you 100% sure, that github actually sent this notification? If so, are you sure, that this notification actually reached the box? If so, are you sure, that this notification actually hit the port the service is listening on? If so, you probably need to investigate, why the service didn't do, what you expected. When you have found the problem either fix it or create a service request/ticket on the jenkins site and wait [forever] or hire a company to solve it. If you haven't done all this so far, the problem is probably not that big, that it is worth to throw that much time/resources on it ...

Wrt. travis: Caching - good to know. When I've dealt with it, it didn't seem to have such a feature, but may be because trusty (not to mention xenial) was not supported at this time, just some beta stuff ...

Anyway, if just "time for tests" is the problem, and travis is otherwise good enough, why not partition the tests, run them in parallel or in consecutive jobs ?

**Comment by eugenk** [comment]
> by triggering an update manually via e-mail, web, etc.) 

The service (not the server) was down and needed to be restarted manually. Only after that, it was possible to trigger a build manually. I did this many times with the old Ontohub. I think it's a problem with the Docker setup and maybe it's best to do this without docker and run Jenkins on an OS that Ontohub/Hets will run on in the future.

> Furthermore I'm pretty sure, that git hub notification service is not a reliable but fire-and-forget service

That's exactly what it is. If the recipient of the web hook is offline, nothing will happen. I'm 100% sure that the notification was sent, but the Jenkins service did not receive it because it was down.

> Anyway, if just "time for tests" is the problem, and Travis is otherwise good enough, why not partition the tests, run them in parallel or in consecutive jobs ?

Travis doesn't support this. It's only one job, one thread and only Ubuntu trusty (and macOS, but we don't need that).

**Comment by jelmd** [comment]
Ah, so you are referring not to ohj... but to john|$a_bremeninstance with "old Ontohub". THan relax, not sure, who set them up/maintained them, but they are history/gone ;-) .

BTW: Wondered, whether github gives some [stats](https://github.com/ontohub/ontohub/settings/hooks/8719224), and indeed: since ohj went online (~15 month agos),  there was one service timeout on 2016-06-20 19:12:11 (probably a config tuning or restart or other issue after being alive for 11 days). Next time it seems that github had some connection problems between 2017-02-20..2017-02-28. Because we didn't change anything on the box/zone/network since setup IIRC, the cause for it was probably on the github site (or at least out of our control). That's it.

So, for me, ohj looks very reliable: 1/641 =~ 99.8% (even with assumed github problems 21/641 =~ 96.7%). There is no need to setup another instance (and yepp, docker is crap (at least for all our scenarios @IKS ) and therefore we do not use docker but LXC zones ;-) ).

Anyway, if you want travis (for whatever reason): One is allowed to submit one job at a time, only? Why not batching 2 or 3 and correlate results? If caching works as described above, the overhead would be ~5min/job. And if all tests together take >50min, I guess no one really cares about 5..10 add. minutes ... ;-)

**Comment by tillmo** [comment]
Travis is very conveniently integrated into github. I cannot see how to support batching and correlating results without undue effort. This is wanted: after all, they want to sell their commercial version...

**Comment by eugenk** [comment]
@jelmd, let's try ohj. Could you please set it up to do the same things as Travis? See the [.travis.yml](https://github.com/spechub/Hets/blob/724f8096df466dc6c7b073363374683625d80a32/.travis.yml). I already set up a Jenkins-project, and left out the "Build" section of the configuration. Most of the configuration is copied over from the "ontohub-PRs" project.

Why have you outsourced the whole test script in the Ontohub instance on ohj to a shell script (file) instead of pasting it directly into the text field (see configuration of "ontohub-PRs" at "Build")?

**Comment by jelmd** [comment]
Hmm, what do you mean with "same things as Travis"? If you want travis, why not use travis ? Not sure, why I should re-invent the wheel ;-)

Wrt. "outsourced the whole test script": I could answer: Because I'm an experienced admin/developer :).
But just in case this you consider it not being sufficient, because:
a) add the answer you would give to the question, why you use git for maintaining hets sources
b) easier testing, troubleshooting, syntax validation, writing, maintenance
c) stability, reliability, flexibility

Finally, don't wonder, but this will not be the last one. As soon as we switch over to jenkins pipelines (and we definitely will - it's just a question of how fast I get the time to study/test/adapt it), another file will join the party: `Jenkinsfile`  ...

**Comment by eugenk** [comment]
> If you want travis, why not use travis ?

See my first post on this issue.

We want to do at least the same things in Jenkins that we do in Travis because Travis has too small limits for the future of Hets's test suite. This probably will diverge, but for now, let's do exactly the same on both platforms. The script that is called by Jenkins should be under version control in this Hets repository.


Wrt. "outsourced script": "To have it under version control" would have sufficed. I can't see the directory structure there. I am just curious.

**Comment by jelmd** [comment]
@tillmo At least travis FAQ says "A build has 120 minutes to run. Should your build take longer, you can split it up using a build matrix.". So my guess is, that it is already supported, just a matter of reading the docs ...

@eugenk define "do exactly the same". I certainly do not waste my time for writing yet another Travis =8-|  .
The script is part of our zone install repo.

**Comment by eugenk** [comment]
> do exactly the same

replicate the `.travis.yml` into a script

**Comment by eugenk** [comment]
> Travis CI [...] will stop the build and add an error message to the build log in the following situations
> 
> * A job produces no log output for 10 minutes
> * A job on travis-ci.org takes longer than 50 minutes
> * A job running on OS X infrastructure takes longer than 50 minutes (applies to travis-ci.org or travis-ci.com)
> * A job on Linux infrastructure on travis-ci.com takes longer than 120 minutes


https://docs.travis-ci.com/user/customizing-the-build#Build-Timeouts

We are using the free travis-ci.org which timeouts after 50 minutes.

**Comment by jelmd** [comment]
And the framed box below says:
> There is no timeout for a build; a build will run as long as all the jobs do as long as each job does not timeout.

**Comment by jelmd** [comment]
The script would be:
```
#!/bin/sh
cat .travis.yml
```
;-)

**Comment by tillmo** [comment]
Interesting, there seems to be no timeout for the build. @eugenk could we split the build into several jobs?

**Comment by eugenk** [comment]
I can only split this into these two jobs:
* compile and run `make check`
* compile, install and run the hets-lib script.

**Comment by tillmo** [comment]
we could split hets-lib into several parts.

**Comment by eugenk** [comment]
Implemented in #1741

**Comment by jelmd** [comment]
Obsolete. We use github actions now.
